app/services/exchange_deleter.rb

```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+class ExchangeDeleter
+  def delete(child, parent)
+    active_upcoming_order_cycles = OrderCycle.active.pluck(:id) + OrderCycle.upcoming.pluck(:id)
+    main_exchanges = Exchange.
+      joins(:order_cycle).
+      where(order_cycles: { id: active_upcoming_order_cycles }).
+      where(sender: parent)
+    parent_variants = ExchangeVariant.where(exchange: main_exchanges.pluck(:id))
+    internal_variants = ExchangeVariant.
+      joins(:exchange).
+      where(
+        exchange_variants: { variant_id: parent_variants.pluck(:variant_id) },
+        exchanges: { sender: child, receiver: child }
+      )
```

Comment:
These are the outgoing variants, right? Maybe if an incoming Exchange is destroyed, it should destroy any related outgoing Exchange. And each Exchange can destroy its related ExchangeVariants. That spreads the code a bit but sounds more robust to me. And each part would be much easier to understand.

Reply:
I don't think this is possible. When you create an order cycle it creates an incoming exchange for each enterprise it is distributing for. Each of those has its own variants of course. It then creates an 'internal' outgoing exchange that combines all of those variants. This internal exchange is used for processing fees (I think). If you destroy the 'internal' outgoing exchange you would essentially destroy the whole order cycle since it would destroy all variants even ones not connected to the incoming exchange that was destroyed.

You have to destroy only the variants linked to the incoming exchange and keep the 'internal' outgoing exchange intact.

Reply:
Now that I rethink this. In the exchange model, I can just destroy the variants from the internal exchange when an incoming exchange is destroyed, leaving the internal exchange itself in place.  Maybe that is what you meant.

Reply:
Actually I ended up pushing it down into the `ExchangeVariant` model. Let me know if I should add anymore tests.

Reply:
Thank you for the detailed explanation. Even after many years of working with OFN code I still remember only my most recent work and forget the details of other parts of the code base. Your explanation sounds solid and I'll look at your new commits.